### ISSUE TITLE
Comments that end up in OpenAPI descriptions should be CommonMark

### ISSUE THREAD

**Issue by sftim** [kubernetes/kubernetes#125020]
### What would you like to be added?

The `description` field is already defined as CommonMark. Now, all strings are acceptable as CommonMark, but this issue is asking for the strings to make obvious sense as CommonMark.

We could also add metadata to comments that are not yet useful CommonMark, and implement a simplification stage that makes such a comment into more renderable CommonMark during OpenAPI generation.

### Why is this needed?

At the moment, consumers of the OpenAPI - for example, generator code that updates https://kubernetes.io/docs/ have to use heuristics to work out whether descriptions are plain text, Markdown, or some hybrid of the two.

---

#### _original issue text was_

~For the OpenAPI specification, add a new field to any definition where a Markdown field description is available.~

- ~this means annotating the doc comments to indicate that they are Markdown~
- ~the `description` OpenAPI field can then either contain the Markdown source, which is kind-of human readable, or it can contain the Markdown transformed into plain text~

**Comment by sftim** [comment]
https://github.com/kubernetes/website/issues/35712#issuecomment-2122467261 inspired this feature request.

/wg api-expression

**Comment by sftim** [comment]
Issues with some (perhaps small) link to this one include:
- https://github.com/kubernetes/kubernetes/issues/10966
- https://github.com/kubernetes/kubernetes/issues/50292
- https://github.com/kubernetes/kubernetes/issues/9097
- https://github.com/kubernetes-sigs/reference-docs/issues/188
- https://github.com/kubernetes/kubernetes/issues/8341
- https://github.com/kubernetes-sigs/reference-docs/issues/293

**Comment by sftim** [comment]
I'm happy to contribute to the outcome but don't have capacity to lead on it.

**Comment by mauri870** [comment]
What would be the proposed name of the field? Can it perhaps be just a boolean flag like `x-is-markdown` and then whoever is parsing the spec can decide if it wants special treatment for markdown content?

**Comment by mauri870** [comment]
Additionally, we need to implement this in a backward-compatible manner. The reference documentation would be a good first step; the ecosystem can then follow suit.

**Comment by sftim** [comment]
> What would be the proposed name of the field?

(no opinion)

--- 
> Can it perhaps be just a boolean flag like x-is-markdown and then whoever is parsing the spec can decide if it wants special treatment for markdown content?

That's trickier; (I just noticed that) the [semantics](https://swagger.io/specification/#rich-text-formatting) of `description` are that it uses CommonMark. If we want to mark places where the doc comments are _not_ valid CommonMark, perhaps we should use `x-invalid-as-markdown`?

Of course, plain text is pretty much always valid Markdown.

Alternatives:
- lint all Kubernetes comments that could become OpenAPI `description` fields; warn on Markdown problems
- mark a subset of Kubernetes code comments as Markdown; lint those as Markdown; lint all other doc comments to make sure they look like plain text and not Markdown

**Comment by sftim** [comment]
I am not sure how we should revise this issue given my new (to me) discovery that `description` is CommonMark.
IMO, either:

- the API reference generators can assume that `description` is CommonMark, and we can drop heuristics

or

- we ought to change things so that `description` is CommonMark, and _then_ we can drop heuristics

@tengqm, do you have an opinion?

**Comment by sftim** [comment]
See https://github.com/kubernetes/kubernetes/issues/125020#issuecomment-2143593987 - we need to rethink the issue description.

**Comment by mauri870** [comment]
Given that the OpenAPI specification states all description fields can be interpreted as CommonMark, I believe any client can confidently render these fields as Markdown and it would be acting according to the spec.

**Comment by sftim** [comment]
I think this should be about the doc comments, then, not about the OpenAPI.

/retitle Comments that end up in OpenAPI descriptions should be CommonMark

**Comment by sftim** [comment]
I'm not sure how we'd frame this. Do we see the cleanup work really as a series of bug fixes? Is this somehow an enhancement or retrospective enhancement that we're making? Something else?

Thoughts welcome.

**Comment by mauri870** [comment]
I think it is more of a visual improvement to the documentation pages. Simply flipping the switch and make them render markdown could make some bugs to surface, but again, today it is not even rendered so I'm fine with the added risk.

For code, we should also enforce somehow that doc comments either use markdown or plain text, no weird formatting in between.

**Comment by sftim** [comment]
> I think it is more of a visual improvement to the documentation pages. Simply flipping the switch and make them render markdown could make some bugs to surface, but again, today it is not even rendered so I'm fine with the added risk.

We already render them as Markdown, but see https://github.com/kubernetes/website/issues/46476 for an example of where that doesn't go how we'd like.

**Comment by sftim** [comment]
I don't think this quite merits the bug label, but it's almost there. OpenAPI says CommonMark and we're not always producing useful CommonMark `description` fields.

**Comment by Jefftree** [comment]
do we know if the limitation is from the generator or the description themselves? I would probably prefer to change descriptions to be commonmark compatible, though it feels like there are cases of valid markdown being rendered improperly?

I haven't looked too deeply at the generation, but just looking at one of the more common issues (Quantity): the description seems to have the proper syntax for markdown (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go) yet the rendering clearly has some issues: https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/
